### ISSUE TITLE
eclair seems to hang onto an HTLC w/o cancelling back

### ISSUE THREAD

**Issue by Roasbeef** [ACINQ/eclair#418]
I have a local `lnd` node with a _single_ channel to starblocks. When I attempt to pay an invoice on `starblocks`, I see the payment go through the first node (: 
```
2018-01-28 15:34:12.840 [DBG] PEER: Sending UpdateAddHTLC(chan_id=3ffb873b22456fa610c007b9d347428ebddb80ae9fd68d5287b38ba1f9b091e4, id=6, amt=1210120 mSAT, expiry=1261114, hash=43784248fa9ae3de399217c74b02cdf4c5102c4cfdceeb76b26ea1126a3b1785) to X:9735
2018-01-28 15:34:12.876 [DBG] PEER: Sending CommitSig(chan_id=3ffb873b22456fa610c007b9d347428ebddb80ae9fd68d5287b38ba1f9b091e4, num_htlcs=0) to D:9735
2018-01-28 15:34:13.089 [DBG] PEER: Received RevokeAndAck(chan_id=3ffb873b22456fa610c007b9d347428ebddb80ae9fd68d5287b38ba1f9b091e4, rev=64770f008c33e8c9d7ff4920950d8f4c575251a1b70f1c26dc26427389fa89f5, next_point=02bfd5ae346a715fe2c3149c25477ea11c15444381ec5730562453dc7573b2d6dc) from D:9735
2018-01-28 15:34:13.101 [DBG] PEER: Received CommitSig(chan_id=3ffb873b22456fa610c007b9d347428ebddb80ae9fd68d5287b38ba1f9b091e4, num_htlcs=0) from X:9735
2018-01-28 15:34:13.103 [DBG] PEER: Sending RevokeAndAck(chan_id=3ffb873b22456fa610c007b9d347428ebddb80ae9fd68d5287b38ba1f9b091e4, rev=648600e1b1c91bfa926e9c23876578f7025fb345c874922eb761c5f704a302cf, next_point=02c947a429e97822ebc831e350db898ec8b20db60a91277bbbec68b25fdd028a91) to D:9735
```

I have a single channel open:
```
i  lncli listchannels
{
    "channels": [
        {
            "active": true,
            "remote_pubkey": "03933884aaf1d6b108397e5efe5c86bcf2d8ca8d2f700eda99db9214fc2712b134",
            "channel_point": "e491b0f9a18bb387528dd69fae80dbbd8e4247d3b907c010a66f45223b87fb3f:0",
            "chan_id": "1386055353091817472",
            "capacity": "15000000",
            "local_balance": "14919139",
            "remote_balance": "4830",
            "commit_fee": "76031",
            "commit_weight": "724",
            "fee_per_kw": "100000",
            "unsettled_balance": "0",
            "total_satoshis_sent": "4830",
            "total_satoshis_received": "0",
            "num_updates": "11",
            "pending_htlcs": [
                {
                    "incoming": false,
                    "amount": "1210",
                    "hash_lock": "Q3hCSPqa4945khfHSwLN9MUQLEz9zut2sm6hEmo7F4U=",
                    "expiration_height": 1261114
                },
                {
                    "incoming": false,
                    "amount": "1210",
                    "hash_lock": "Q3hCSPqa4945khfHSwLN9MUQLEz9zut2sm6hEmo7F4U=",
                    "expiration_height": 1261114
                },
                {
                    "incoming": false,
                    "amount": "1210",
                    "hash_lock": "Q3hCSPqa4945khfHSwLN9MUQLEz9zut2sm6hEmo7F4U=",
                    "expiration_height": 1261114
                }
            ],
            "csv_delay": 144
        }
    ]
}
```

Observe the the HTLC's are just left there pending, and at this rate, will only expire in a day or so. 

The channel is with endurance: 
```
⛰i  lncli getnodeinfo 03933884aaf1d6b108397e5efe5c86bcf2d8ca8d2f700eda99db9214fc2712b134
{
    "node": {
        "last_update": 1517163453,
        "pub_key": "03933884aaf1d6b108397e5efe5c86bcf2d8ca8d2f700eda99db9214fc2712b134",
        "alias": "endurance",
        "addresses": [
            {
                "network": "tcp",
                "addr": "X:9735"
            }
        ],
        "color": "#ffff00"
    },
    "num_channels": 437,
    "total_capacity": "2820108499"
}
```

The HTLC is never cancelled back or settled, and ends up just sitting there on the commitment transaction. I've had some users complain about "stuck" HTLC's, and it seems that starblocks may be the cause.

**Comment by Roasbeef** [comment]
Paying from another node (which doesn't have a direct channel open), I see the same effect.

**Comment by pm47** [comment]
So setup is this:

`lnd` --- `endurance` --- `starblocks`

HTLCs were also stuck in the `endurance` --- `starblocks` channel, it seems `starblocks` just didn't send fulfills. I manually failed the HTLCs so they won't timeout.

It could be related to the fact that we had an automated db maintenance planned tonight (`starblocks` fetches preimages from there), but timing do not seem to match with your logs since the upgrade took place at 29 jan 2018 04:57:35 UTC+1 (assuming your node's tz is PST?).

**Comment by pm47** [comment]
Ok found the bug.

Endurance currently runs the `tcp-pull-mode` experimental branch which was buggy: some messages were never tcp-acked which caused htlc to be stucked between endurance and starblocks.

I fixed it with 9b25d55c26f5aa67f0331e7ea4ed61b666245a2f.

Thanks!
